src/Fmt_ast.ml

```diff
@@ -634,25 +634,28 @@ and fmt_core_type c ?(box = true) ?(in_type_declaration = false) ?pro
           | None -> false
           | Some x -> exposed_right_typ x )
       in
-      let force_break =
-        match c.conf.type_decl with
-        | `Sparse when c.conf.space_around_collection_expressions -> true
-        | _ -> false
+      let space_around = c.conf.space_around_collection_expressions in
+      let closing =
+        fits_breaks
+          (if protect_token || space_around then " ]" else "]")
+          ( if Poly.(c.conf.type_decl = `Sparse) && space_around then
+            "@;<1000 0>]"
+          else "@ ]" )
       in
       hvbox 0
-        ( fits_breaks "[" "["
-        $ ( match (flag, lbls, rfs) with
-          | Closed, None, [Rinherit _] -> fmt " | " $ row_fields rfs
-          | Closed, None, _ -> fits_breaks "" " " $ row_fields rfs
-          | Open, None, _ -> fmt "> " $ row_fields rfs
-          | Closed, Some [], _ -> fmt "< " $ row_fields rfs
-          | Closed, Some ls, _ ->
-              fmt "< " $ row_fields rfs $ fmt " > "
-              $ list ls "@ " (fmt "`" >$ str)
-          | Open, Some _, _ -> impossible "not produced by parser" )
-        $ fits_breaks
-            (if protect_token then " ]" else "]")
-            (if force_break then "@;<1000 0>]" else "@ ]") )
+        ( match (flag, lbls, rfs) with
+        | Closed, None, [Rinherit _] ->
+            fmt "[ | " $ row_fields rfs $ closing
+        | Closed, None, _ ->
+            let opening = if space_around then "[ " else "[" in
+            fits_breaks opening "[ " $ row_fields rfs $ closing
+        | Open, None, _ -> fmt "[> " $ row_fields rfs $ closing
```

Comment:
Should we remove those spaces when `space-around-collection-expressions` is `false` ?
eg.
```ocaml
type t = [>`A]
type t = [<`A]
type t = [<`A>`B]
```

Reply:
After writing this, I'd say no.